Classes/TestingFramework.php

```diff
@@ -0,0 +1,104 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+namespace Codappix\Typo3PhpDatasets;
+
+use Doctrine\DBAL\DBALException;
+use InvalidArgumentException;
+
+/**
+ * Only use this within an FunctionalTestCase.
+ */
+trait TestingFramework
+{
+    protected function importPHPDataSet(string $filePath): void
+    {
+        $this->ensureFileExists($filePath);
+
+        $dataSet = include $filePath;
+        try {
+            (new PhpDataSet())->import($dataSet);
+        } catch (DBALException $e) {
+            self::fail('SQL Error for PHP data set "' . $filePath . '":' . PHP_EOL . $e->getMessage());
+        }
+    }
+
+    /**
+     * Highly inspired by TYPO3 testing framework.
+     */
+    protected function assertPHPDataSet(string $filePath): void
+    {
+        $this->ensureFileExists($filePath);
+
+        $dataSet = include $filePath;
+        $failMessages = [];
+
+        foreach ($dataSet as $tableName => $expectedRecords) {
+            $records = $this->getAllRecords($tableName, true);
+
+            foreach ($expectedRecords as $assertion) {
+                $result = $this->assertInRecords($assertion, $records);
+                if ($result === false) {
+                    // Handle error
+                    if (isset($assertion['uid']) && empty($records[$assertion['uid']])) {
+                        $failMessages[] = 'Record "' . $tableName . ':' . $assertion['uid'] . '" not found in database';
+                        continue;
+                    }
+                    if (isset($assertion['uid'])) {
+                        $recordIdentifier = $tableName . ':' . $assertion['uid'];
+                        $additionalInformation = $this->renderRecords($assertion, $records[$assertion['uid']]);
+                    } else {
+                        $recordIdentifier = $tableName;
+                        $additionalInformation = $this->arrayToString($assertion);
+                    }
+
+                    $failMessages[] = 'Assertion in data-set failed for "' . $recordIdentifier . '":' . PHP_EOL . $additionalInformation;
+                    continue;
+                }
+
+                // Unset asserted record
+                unset($records[$result]);
+                // Increase assertion counter
+                self::assertTrue($result !== false);
+            }
+
+            if (!empty($records)) {
+                foreach ($records as $record) {
+                    $recordIdentifier = $tableName . ':' . $record['uid'];
+                    $failMessages[] = 'Not asserted record found for "' . $recordIdentifier . '".';
```

Comment:
Sounds strange

Reply:
That's copied from testing framework. What's strange? There is a not asserted record, sounds fine to me.

Reply:
Ok - let's keep it like it is.